File: Java/TinyScript/src/main/java/lexer/Lexer.java
package lexer;

import common.AlphabetHelper;
import common.PeekIterator;

import java.util.ArrayList;
import java.util.stream.Stream;

public class Lexer {
    /*
     * 词法分析器
     * @params {Stream}source 源代码
     * @return {ArrayList}
     */
    public ArrayList<Token> analyse(Stream source) throws LexicalException {

        var tokens = new ArrayList<Token>();
        var it = new PeekIterator<Character>(source, (char)0);

        while (it.hasNext()) {

            char c = it.next();

            if (c == 0) break;

            char lookahead = it.peek();

            if (c == ' ' || c == '\n') continue;

            // 删除注释
            if (c == '/') {
                if (lookahead == '/') {
                    while (it.hasNext() && (c = it.next()) != '\n');
                    continue;
                } else if (lookahead == '*') {

                    boolean valid = false;

                    while (it.hasNext()) {
                        char p = it.next();

                        if (p == '*' && it.peek() == '/') {
                            it.next();
                            valid = true;
                            break;
                        }
                    }

                    if (!valid) {
                        throw new LexicalException("Comments not match");
                    }

                    continue;
                }
            }

            if (c == '{' || c == '}' || c == '(' || c == ')') {
                tokens.add(new Token(TokenType.BRACKET, c + ""));
                continue;
            }

            if (c == '"' || c == '\'') {
                it.putBack();
                tokens.add(Token.makeString(it));
                continue;
            }

            if (AlphabetHelper.isLetter(c)) {
                it.putBack();
                tokens.add(Token.makeVarOrKeyword(it));
                continue;
            }

            if (AlphabetHelper.isNumber(c)) {
                it.putBack();
                tokens.add(Token.makeNumber(it));
                continue;
            }

            if ((c == '+' || c == '-' || c == '.') && AlphabetHelper.isNumber(lookahead)) {

                var lastToken = tokens.size() == 0 ? null : tokens.get(tokens.size() - 1);

                if (lastToken == null || !lastToken.isNumber() || lastToken.isOperator()) {
                    it.putBack();
                    tokens.add(Token.makeNumber(it));
                    continue;
                }
            }


            if (AlphabetHelper.isOperator(c)) {
                it.putBack();
                tokens.add(Token.makeOp(it));
                continue;
            }

            throw new LexicalException(c);

        } // end while

        return tokens;
    }
}

File: JavaScript/TinyScript/src/lexer/Lexer.js
const PeekIterator      = require('../common/PeekIterator')
const Token             = require('./Token')
const TokenType         = require('./TokenType')
const AlphabetHelper    = require('./AlphabetHelper')
const LexicalException  = require('./LexicalException')

class Lexer {

    analyse(source) {

        const tokens = []
        const it = new PeekIterator(source, '\0')

        while (it.hasNext()) {

            let c = it.next()

            if (c === '\0') break

            const lookahead = it.peek()

            if (c === ' ' || c === '\n' || c === '\r') continue

            if (c === '/') {

                if (lookahead === '/') {
                    while (it.hasNext() && (c = it.next()) !== '\n');
                    continue
                } else if (lookahead === '*') {
                    let vali = false

                    while (it.hasNext()) {
                        const p = it.next()

                        if (p === '*' && it.peek() === '/') {
                            vali = true
                            it.next()
                            break
                        }
                    }

                    if (!vali) {
                        throw new LexicalException('Comments not matched')
                    }

                    continue
                }
            }

            if (c === '{' || c === '}' || c === '(' || c === ')') {
                tokens.push(new Token(TokenType.BRACKET, c))
                continue
            }

            if (c === '"' || c === "'") {
                it.putBack()
                tokens.push(Token.makeString(it))
                continue
            }

            if (AlphabetHelper.isLetter(c)) {
                it.putBack()
                tokens.push(Token.makeVarOrKeyword(it))
                continue
            }

            if (AlphabetHelper.isNumber(c)) {
                it.putBack()
                tokens.push(Token.makeNumber(it))
                continue
            }

            if ((c === '+' || c === '-' || c === '.') && AlphabetHelper.isNumber(lookahead)) {

                const lastToken = tokens[tokens.length - 1] || null

                // 前面没有值 或者前面不是一个变量
                if (lastToken == null || !lastToken.isValue()) {
                    it.putBack()
                    tokens.push(Token.makeNumber(it))
                    continue
                }
            }

            if (AlphabetHelper.isOperator(c)) {
                it.putBack()
                tokens.push(Token.makeOp(it))
                continue
            }

            throw LexicalException.fromChar(c)

        } // end while

        return tokens
    }
}

module.exports = Lexer
